specs/phase1/validator.md

```diff
@@ -0,0 +1,569 @@
+
+##### Custody key reveals
+
+Up to `MAX_CUSTODY_KEY_REVEALS`, [`CustodyKeyReveal`](./custody-game.md#custodykeyreveal) objects can be included in the `block`. The custody key reveals must satisfy the verification conditions found in [custody key reveal processing](./custody-game.md#custody-key-reveals). The validator receives a small reward for each custody key reveal included.
+
+##### Early derived secret reveals
+
+Up to `MAX_EARLY_DERIVED_SECRET_REVEALS`, [`EarlyDerivedSecretReveal`](./custody-game.md#earlyderivedsecretreveal) objects can be included in the `block`. The early derived secret reveals must satisfy the verification conditions found in [early derived secret reveal processing](./custody-game.md#custody-key-reveals). The validator receives a small "whistleblower" reward for each early derived secrete reveal included.
+
+##### Shard transitions
+
+Exactly `MAX_SHARDS` [`ShardTransition`](./beacon-chain#shardtransition) objects are included in the block. Default each to an empty `ShardTransition()`. Then for each committee assigned to the slot with an associated `committee_index` and `shard`, set `shard_transitions[shard] = full_transitions[winning_root]` if the committee had enough weight to form a crosslink this slot.
+
+Specifically:
+* Call `shards, winning_roots = get_shard_winning_roots(state, block.slot, block.body.attestations)`
+* Let `full_transitions` be a dictionary mapping from the `shard_transition_root`s found in `attestations` to the corresponding full `ShardTransition`
+* Then for each `shard` and `winning_root` in `zip(shards, winning_roots)` set `shard_transitions[shard] = full_transitions[winning_root]`
+
+*Note*: The `state` passed into `get_shard_winning_roots` must be transitioned into the epoch of `block.slot` to run accurately due to the internal use of `get_online_validator_indices`.
+
+```python
+def get_shard_winning_roots(state: BeaconState,
+                            slot: Slot,
+                            attestations: Sequence[Attestation]) -> Tuple[Sequence[Shard], Sequence[Root]]:
+    shards = []
+    winning_roots = []
+    online_indices = get_online_validator_indices(state)
+    committee_count = get_committee_count_at_slot(state, slot)
+    for committee_index in map(CommitteeIndex, range(committee_count)):
+        shard = compute_shard_from_committee_index(state, committee_index, slot)
+        # All attestations in the block for this committee/shard and current slot
+        shard_attestations = [
+            attestation for attestation in attestations
+            if attestation.data.index == committee_index and attestation.data.slot == slot
```

Comment:
```suggestion
            if attestation.data.index == committee_index and attestation.data.slot == compute_previous_slot(slot)
```

Shouldn't `beacon_block` at slot N have `on_time_attestations` from slot N - 1 ?

Reply:
Agreed  👍  And you can use `is_on_time_attestation(state, attestation)` helper.
```suggestion
            if is_on_time_attestation(state, attestation) and attestation.data.index == committee_index
```

And simplify the `get_shard_winning_roots` arguments to:
```python
def get_shard_winning_roots(state: BeaconState,
                            attestations: Sequence[Attestation]) -> Tuple[Sequence[Shard], Sequence[Root]]:
    slot = state.slot
    ...